monai/losses/ssim_loss.py

```diff
@@ -0,0 +1,91 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+from monai.utils.type_conversion import convert_to_tensor
+
+
+class SSIMLoss(nn.Module):
+    """
+    Build a Pytorch version of the SSIM loss function based on the original formula of SSIM
+
+    Modified and adopted from:
+        https://github.com/facebookresearch/fastMRI/blob/main/banding_removal/fastmri/ssim_loss_mixin.py
+
+    For more info, visit
+        https://vicuesoft.com/glossary/term/ssim-ms-ssim/
+
+    SSIM reference paper:
+        Wang, Zhou, et al. "Image quality assessment: from error visibility to structural
+        similarity." IEEE transactions on image processing 13.4 (2004): 600-612.
+    """
+
+    def __init__(self, win_size: int = 7, k1: float = 0.01, k2: float = 0.03, spatial_dims: int = 2):
+        """
+        Args:
+            win_size: gaussian weighting window size
+            k1: stability constant used in the luminance denominator
+            k2: stability constant used in the contrast denominator
+            spatial_dims: if 2, input shape is expected to be (B,C,W,H). if 3, it is expected to be (B,C,W,H,D)
+        """
+        super().__init__()
+        self.win_size = win_size
+        self.k1, self.k2 = k1, k2
+        self.spatial_dims = spatial_dims
+        self.register_buffer(
+            "w", torch.ones([1, 1] + [win_size for _ in range(spatial_dims)]) / win_size**spatial_dims
+        )
+        self.cov_norm = (win_size**2) / (win_size**2 - 1)
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor, data_range: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            x: first sample (e.g., the reference image).
```

Comment:
Add hint for the shape of a and y. Thank you!

Reply:
added, thanks!